spec/models/product_spec.rb

```diff
@@ -0,0 +1,87 @@
+require "rails_helper"
+require "active_storage_validations/matchers"
+
+RSpec.describe Product, type: :model do
+  describe "associations" do
+    it { should have_many(:comments).dependent(:destroy) }
+    it { should have_many(:cart_details).dependent(:destroy) }
+    it { should belong_to(:category).optional }
+    it { should have_one_attached(:image) }
+  end
+
+  describe "validations" do
+    context "when validations pass" do
+      let(:product) { build(:product) }
+
+      it "validates presence of name" do
+        expect(product).to validate_presence_of(:name)
+      end
+
+      it "validates presence of description" do
+        expect(product).to validate_presence_of(:description)
+      end
+
+      it "validates presence of price" do
+        expect(product).to validate_presence_of(:price)
+      end
+
+      it "validates numericality of price" do
+        expect(product).to validate_numericality_of(:price).is_greater_than(0)
+      end
+
+      it "validates presence of remain quantity" do
+        expect(product).to validate_presence_of(:remain_quantity)
+      end
+
+      it "validates numericality of remain quantity" do
+        expect(product).to validate_numericality_of(:remain_quantity).only_integer.is_greater_than_or_equal_to(0)
+      end
+
+      it "validates presence of category_id" do
+        expect(product).to validate_presence_of(:category_id)
+      end
+
+      it "validates image content type" do
+        expect(product).to validate_content_type_of(:image).allowing("image/png", "image/jpeg")
+      end
+
+      it "validates image size" do
+        expect(product).to validate_size_of(:image).less_than(Settings.max_image_data.megabytes)
+      end
+    end
+
+    context "when validations fail" do
+      let(:product) { build(:product, price: -10, remain_quantity: -5) }
+
+      it "fails to validate numericality of price" do
+        expect(product).not_to be_valid
+        expect(product.errors[:price]).to include(I18n.t("errors.messages.greater_than", count: 0))
+      end
+
+      it "fails to validate numericality of remain quantity" do
+        expect(product).not_to be_valid
+        expect(product.errors[:remain_quantity]).to include(I18n.t("errors.messages.greater_than_or_equal_to", count: 0))
+      end
+
+      it "fails to validate image content type" do
+        product.image.attach(io: File.open("spec/fixtures/files/test.txt"), filename: "test.txt", content_type: "text/plain")
+        expect(product).not_to be_valid
+        expect(product.errors[:image]).to include(I18n.t("errors.messages.content_type_invalid"))
+      end
+    end
+  end
+
+  describe "scopes" do
+    let!(:category) { create(:category) }
```

Comment:
let vs let! khác nhau như nào em

Reply:
Dạ,
Khi sử dụng let, biến này sẽ chỉ được khởi tạo (hoặc tính toán) khi lần đầu tiên nó được gọi trong một example. Điều này có nghĩa là nếu không gọi biến đó trong một example, nó sẽ không bao giờ được khởi tạo.
 Ngược lại, let! sẽ khởi tạo (hoặc tính toán) biến ngay lập tức trước khi bất kỳ example nào chạy. Điều này đảm bảo rằng biến đó luôn được khởi tạo, bất kể nó có được gọi hay không